OSDOCS-930 - Include section on the OpenStack Manila CSI Driver

bobfuru opened the pull request:
Work being tracked in [OSDOCS-930](https://issues.redhat.com/browse/OSDOCS-930) and [OSDOCS-1117](https://issues.redhat.com/browse/OSDOCS-1117).

**Storage -> Using CSI -> Using CSI OpenStack Manila**
Preview link: https://osdocs-930-manila--ocpdocs.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-manila.html

bobfuru commented:
@duanwei33 PTAL?

duanwei33 commented:
@bobfuru Sorry for the delay, I think there is still some update for this feature, I list some items not align with our test:
1. Project name should be manila-csi instead of csi-manila
2. After "Installing the CSI Manila Operator" and before "Dynamically provisioning CSI Manila volumes", there is one more step to install ManilaDriver, which will create csi driver related resource and storage class (It is more like install LocalVolume in local storage), it can be installed from web or client.
3. In "Dynamically provisioning CSI Manila volumes", the storage class name for manila should be "csi-manila-ceph" instead of "csi-manila-nfs" 
4. Just suggest only keep command line way in  "Dynamically provisioning CSI Manila volumes" to simple the doc (Just a suggestion)
5. "pvc-clone.yaml" should be updated to "pvc-manila.yaml".
6. Not very sure the note for RWO: 
> Use RWO mode if you want to prevent additional Pods from being dynamically provisioned.

I can provide env with manila operator if you needed.
BTW, just wondering when the doc should be finished reviewed by QE, I know the GA date is June 18th?

duanwei33 commented:
Hi Bob,
I checked some pr, seems something changed like namespace, install mode, driver name. 
I will give another review next week after verified ON_QA bugs.

bobfuru commented:
Hi Wei,

Thanks for the feedback. See my inline comments below...

> @bobfuru Sorry for the delay, I think there is still some update for this feature, I list some items not align with our test:

That makes sense. Because I didn't have CSI Manila Operator in OLM, I was mostly guessing. :)

> 1. Project name should be manila-csi instead of csi-manila
OK, changed. Do you think we need this step? I think a default namespace is created if we leave this out.

> 2. After "Installing the CSI Manila Operator" and before "Dynamically provisioning CSI Manila volumes", there is one more step to install ManilaDriver, which will create csi driver related resource and storage class (It is more like install LocalVolume in local storage), it can be installed from web or client.

I added a section for instructions about installing ManilaDriver (or whatever name it will be called). So the Operator doesn't deploy the driver? I thought that's one of the features of CSI Operators? I'm not sure what the steps are to do this, so please fill in/comment on them as you have more details.

> 3. In "Dynamically provisioning CSI Manila volumes", the storage class name for manila should be "csi-manila-ceph" instead of "csi-manila-nfs"

Fixed. 

> 4. Just suggest only keep command line way in  "Dynamically provisioning CSI Manila volumes" to simple the doc (Just a suggestion)

Because we have console instructions, I'm inclined to keep those here, along with CLI. 

> 5. "pvc-clone.yaml" should be updated to "pvc-manila.yaml".

Fixed!

> 6. Not very sure the note for RWO:
> 
> > Use RWO mode if you want to prevent additional Pods from being dynamically provisioned.

This came from a BZ enhancement for local storage Operator install, so I thought it would also be useful context here. If you think it's incorrect, please let me know and I'll remove.

> I can provide env with manila operator if you needed.

Yes, that would be most helpful, thanks!

> BTW, just wondering when the doc should be finished reviewed by QE, I know the GA date is June 18th?

As of now, our Docs freeze date is June 8, but I know GA is now July 9, so we will probably have additional time for this doc to be finished with QE review. Or is June 8 even possible for you?

bobfuru commented:
> Hi Bob,
> I checked some pr, seems something changed like namespace, install mode, driver name.
> I will give another review next week after verified ON_QA bugs.

Perfect, thank you, @duanwei33 !! I did what I could for now based on the info I have (and no test env). Appreciate you looking now and again next week!!

bobfuru commented:
@jeana-redhat PTAL for clarity, typos, links, etc. Thanks!

jeana-redhat commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,88 @@
+  storageClassName: csi-manila-nfs <2>
+----
++
+<1> Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster. Use RWO mode to prevent additional Pods from being dynamically provisioned.
```
Can we assume any user reading this knows the ReadWriteMany:RWX abbreviation and others like it?

bobfuru replied:
Not necessarily, but we don't generally write out RWX/ROX/RWO except in the access modes section of Storage docs. IMO it's easy to search for the answer if a newer user isn't familiar. But I'm open to revisiting for standardization, esp. since we try to spell out other storage initialisms, such as PV, PVC, and CSI, on first instance. ¯\_(ツ)_/¯

jeana-redhat commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,88 @@
+<1> Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster. Use RWO mode to prevent additional Pods from being dynamically provisioned.
+<2> The name of the storage class that provisions the storage back end. This was created by the cluster administrator.
++
+. Create the object you saved in the previous step by running the following command:
```
```suggestion
. To create the object that you saved in the previous step, run the following command:
```

jboxman replied:
@jeana-redhat, I've been using this construction, but I'm curious why _To ... run_ is preferable to _Create ... by running_; My original background is not technical writing, so it may be obvious.

jeana-redhat replied:
Disclaimer: I am neither an editor nor an English major, so I may have terms wrong, but will do my best ;)
This suggestion is based on a combination of favoring goal/purpose-oriented statements ("**To accomplish X**, do Y") and using the imperative to describe the action the user should take ("To accomplish X, **do Y**"). "Accomplish X by doing Y" is goal-oriented, but the imperative is on the goal rather than the actual user action. 
Personally, I don't feel there's a huge difference in meaning here, but if you look at the ISG it uses this form consistently (there's an example of this change being explicitly recommended at the end of p 225/start of p 226). There is also of course the additional benefit to translation if we all always use the same forms.

bobfuru replied:
My 2¢: 

Here's another case where minimalism, l10n, and ISG form concentric circles around doing what's best for the user, though with conflicting guidance. 

In this module, the primary goal is stated in the intro: "To dynamically create a CSI Manila volume using the command-line interface...." And I've been habituated to front-load proc sentences with verbs/actions, which is what I see most frequently in our docs ("Click Create to create the instance" instead of "To create the instance, click Create", for example).

Another issue I see with favoring goal-oriented statements is the risk of extraneous info ("Click Save" vs. "To save (?) the file, click Save"). 

Also, I've been forcing myself to use ISG and add "by running the following command(s)", so I'll add that here.

jeana-redhat replied:
Thanks! I of course haven't read enough of our docs to have a feel for what's really common yet :) But consistency is a good thing to add to the balance!

jeana-redhat commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,88 @@
++
+A new PVC is created.
+
+. Verify that the volume was created and is ready by running the following command:
```
```suggestion
. To verify that the volume was created and is ready, run the following command:
```

bobfuru replied:
I agree with your previous comment that there's not much difference here. I like your construction better here ("run the following command" sounds better than "by running"), so I'll update this line.

jeana-redhat commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,88 @@
++
+The `pvc-manila` shows that it is `Bound`.
+
+You are now ready to use the newly created PVC to configure a Pod.
```
```suggestion
You can now use the new PVC to configure a Pod.
```

jboxman commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,88 @@
+
+. Define the size of the storage claim.
+
+. Click *Create* to create the PersistentVolumeClaim and generate a PersistentVolume.
```
@jeana-redhat, so are we, honestly.

jboxman commented:
@jeana-redhat, I don't usually drive-by comment, but this is impressive peer review work and I wanted to say so!

bobfuru commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,88 @@
+
+* Storage must exist in the underlying infrastructure before it can be mounted as a volume in {product-title}.
+
+.UI procedure
```
I changed it to `Procedure (UI)` and later `Procedure (CLI)` as that's the convention I've seen adopted elsewhere in our docs (e.g., https://docs.openshift.com/container-platform/4.4/storage/persistent_storage/persistent-storage-efs.html#efs-pvc_persistent-storage-efs)

jeana-redhat replied:
Aha! Thanks for letting me know what's used throughout :)

bobfuru commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,25 @@
+[id="persistent-storage-csi-manila-storage-class_{context}"]
+= Installing the CSI Manila driver
+
+Before PersistentVolumeClaims can be created, you must install the CSI ManilaDriver, which creates the related the CSI Manila driver resource and StorageClass that must exist in the {product-title} cluster. The following instructions create the ManilaDriver needed for dynamic provisioning.
```
Yeah, it's confusing...we are sometimes writing the resource type and other times the instance/name. I have updated based on your suggestions to be clearer (hopefully). 😃

bobfuru commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,25 @@
+
+.Procedure
+
+. Install the CSI ManilaDriver from the web console.
```
I refactored this procedure a bit.

bobfuru commented on storage/container_storage_interface/persistent-storage-csi-manila.adoc:
```diff
@@ -0,0 +1,26 @@
+
+== Overview
+
+{product-title} is capable of provisioning PVs using the CSI driver for the link:https://wiki.openstack.org/wiki/Manila[OpenStack Manila] shared file system service.
```
Yeah, I missed this one...good catch. :)

bobfuru commented:
@Fedosin @tombarron @jsafrane PTAL - While I can't verify the OperatorHub install yet, I'd like to know if I've captured correct details (namespace, for example). For background info, I used these CSI Manila docs:
https://docs.google.com/document/d/1h3vmgvM0m7VACppw_TZAgfhv4wbXjfx9MkT833g0-mI/edit#
https://github.com/openshift/csi-driver-manila-operator

bobfuru commented:
@duanwei33 I've made updates throughout, PTAL, thanks.

jsafrane commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,48 @@
+
+.Procedure
+
+To create the `manila-csi` project from the web console, follow these steps:
```
You refer to the project as `openshift-csi-driver-manila-operator` below. IMO, `openshift-csi-driver-manila-operator` is the correct one.

bobfuru replied:
That was my thinking, too, but was matching what was in the deployment GDoc. Fixed.

jsafrane commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,48 @@
+
+.Procedure
+
+To create the `manila-csi` project from the web console, follow these steps:
+
+. Log in to the {product-title} web console.
+
+. Navigate to *Projects*.
+
+. Click *Create Project*.
+
+. To create the installed namespace, enter *openshift-csi-driver-manila-operator* in the *Name* field.
+
+. Edit the fields for the *Display Name* and *Description* using the values that you want.
+
+. Click *Create*.
```
IMO, project / namespace creation is not necessary. OLM will create the namespace automatically when user follows this:

> On the *Install Operator* page, be sure that *A specific namespace on the cluster* is selected. Select *openshift-csi-driver-manila-operator* from the drop-down menu.

bobfuru replied:
Makes sense. I wasn't 100% sure about that as I couldn't test in 4.5 yet. I've removed the steps to create project/namespace.

jsafrane commented:
lgtm

duanwei33 commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,34 @@
+
+All of the driver's resources are created in the `openshift-manila-csi-driver` namespace.
+
+Once finished, the OpenStack Manila CSI Driver Operator is listed in the *Installed Operators* section of the web console.
```
After that, we need to create ManilaDriver instance to create manila driver, this will create all related components including storage class "csi-manila-ceph", which we can use it for dynamic provision.

bobfuru replied:
I was under the impression that this Operator creates the Manila driver and thus a default storage class, similar to AWS EBS Operator: https://github.com/openshift/openshift-docs/pull/22350/files#diff-8e09bc6a796983013e07888f275920e5R15
If that's not the case, can you describe the steps to create the manila driver, please?

duanwei33 replied:
@bobfuru Yes, it should be a similar step with AWS EBS CSI Operator, but looks it also missed the step to create CSI driver and related resource. If I understand well, actually it is also like the local volume in https://docs.openshift.com/container-platform/4.4/storage/persistent_storage/persistent-storage-local.html#local-volume-cr_persistent-storage-local

Following is install ManilaDriver from UI/CLI, and check the installation for your rreference.
Install ManilaDriver from the UI
1. On the "Installed Operators" page, select and entry "Manila CSI Driver Operator", in the "ManilaDriver" tab, you can see installed ManilaDrivers here.
2. Click "Create ManilaDriver" button, then click "Create" button to create manila driver. please note in "YAML View", the name "cluster" cound not be renamed, and it doesn't support specify a certain namespace.

Install ManilaDriver from the Click
1. Create the ManilaDriver  
Example: ManilaDriver
$ more maniladriver.yaml
apiVersion: csi.openshift.io/v1alpha1
kind: ManilaDriver
metadata:
  name: cluster
  
 please note the name "cluster" cound not be renamed, also it doesn't support specify a certain namespace. 
 
2. Create the ManilaDriver in your OpenShift Container Platform cluster, specifying the file you just created:
$ oc create -f maniladriver.yaml


Verify ManilaDriver installed successfully
1. Verify corresponding resource are created successfully:
$ oc get all -n openshift-manila-csi-driver
NAME                                                         READY   STATUS    RESTARTS   AGE
pod/csi-nodeplugin-nfsplugin-lzvpm                           1/1     Running   0          18h
pod/csi-nodeplugin-nfsplugin-slvg2                           1/1     Running   0          18h
pod/csi-nodeplugin-nfsplugin-xmps9                           1/1     Running   0          18h
pod/openstack-manila-csi-controllerplugin-7d4f5d985b-mw4x5   3/3     Running   0          17h
pod/openstack-manila-csi-nodeplugin-6xchs                    2/2     Running   0          18h
pod/openstack-manila-csi-nodeplugin-bkcmz                    2/2     Running   0          18h
pod/openstack-manila-csi-nodeplugin-rlpps                    2/2     Running   0          18h

NAME                                             DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
daemonset.apps/csi-nodeplugin-nfsplugin          3         3         3       3            3           <none>          18h
daemonset.apps/openstack-manila-csi-nodeplugin   3         3         3       3            3           <none>          18h

NAME                                                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/openstack-manila-csi-controllerplugin   1/1     1            1           18h

NAME                                                               DESIRED   CURRENT   READY   AGE
replicaset.apps/openstack-manila-csi-controllerplugin-7d4f5d985b   1         1         1       17h

2. Verify storage class created successfully:
$ oc get storageclass csi-manila-ceph
NAME              PROVISIONER                RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
csi-manila-ceph   manila.csi.openstack.org   Delete          Immediate           false                  18h

bobfuru replied:
Thank you, @duanwei33 - I have some formatting clean-up still to do, but hoping you can take a look at my updates based on your awesome help with this. 😃

duanwei33 replied:
@bobfuru The persistent-storage-csi-manila-install-driver looks good to me.
Feel free to update my draft  :)

duanwei33 replied:
@bobfuru As we decide to use "Manila CSI Driver Operator" as the official name, so I suggest using, I see somewhere need to be updated, WDYT? 
- "Manila CSI Driver Operator" instead of "CSI Driver Manila Operator"
- "Manila CSI Driver " instead of "CSI Manila Driver"
- "manila csi driver" instead of "csi manila driver" or "manila driver" when talking about the csi driver

bobfuru replied:
Thanks, @duanwei33 - It's been tricky getting the naming right but I think I have updated everything as you indicated. But the name shows up differently in console/Operator Hub: `CSI Driver Manila Operator`. In some places, when talking about the resource, I went with `ManilaDriver` rather than `Manila CSI Driver`.

bobfuru commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,34 @@
+
+. Click *Install*.
+
+. On the *Install Operator* page, be sure that *A specific namespace on the cluster* is selected. Select *openshift-csi-driver-manila-operator* from the drop-down menu.
```
I've updated docs to match web console workflow:
> `. On the *Install Operator* page, select *openshift-manila-csi-driver-operator* from the *Installed Namespace* drop-down menu.`

bobfuru commented:
@Fedosin Bump - can you PTAL and LGTM so that I can merge?

bobfuru commented:
> I left some remarks inline.
> 
> Would it be possible to see a rendered version of the entire document with these proposed changes in context? That would make it much easier to compare with the information provided for other storage drivers to see if there's some topic we're missing.

@tombarron Thank you for your feedback! I'll apply your comments and share an updated preview build with you early next week.

bobfuru commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+  resources:
+    requests:
+  	storage: 10Gi
+  storageClassName: slow <2>
```
Thanks for pointing this out. I received QE feedback to match the source doc I was using, but I too questioned it. I'll change to `gold`.

bobfuru commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+
+.Prerequisites
+
+* Storage must exist in the underlying infrastructure before it can be mounted as a volume in {product-title}.
```
Yes, that makes more sense.

bobfuru commented on modules/persistent-storage-csi-manila-install-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+
+.Prerequisites
+* You have access to the {product-title} web console.
+* NFS share types are used for provisioning.
```
Not entirely sure either, as this prereq was picked up from preexisting source content. But your wording makes better sense.

Fedosin commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+[id="persistent-storage-csi-manila-dynamic-provisioning_{context}"]
+= Dynamically provisioning Manila CSI volumes
+
+{product-title} installs a default StorageClass.
```
This phrase is little bit confusing for me... In OpenShift on OpenStack the default StorageClass `standard` is for Cinder, and it is created by cluster-storage-operator: https://github.com/openshift/cluster-storage-operator/blob/665c4647a815489171eb3f9eff149ae714c98e41/assets/openstack.yaml#L6
Then, there could be several optional StorageClasses for Manila, one for each share type: https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-manila-csi-plugin.md#controller-service-volume-parameters
If I understood it correctly, I suggest to write this line with something like `{product-title} installs a StorageClass for each available Manila share type.`.

bobfuru replied:
This makes more sense, updated.

Fedosin commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+.Prerequisites
+
+* OpenStack is deployed with appropriate Manila share infrastructure so that it can be used to dynamically provision and mount volumes in {product-title}.
+
```
There is another prerequisite, that all OCP compute nodes should be connected to the storage network to be able to mount volumes from this network. 
Actually, it's an obvious prerequisite, I don't know if other platforms mention it in their documentation.

bobfuru replied:
We don't mention it elsewhere and our guidance is not to include obvious prereqs, so I'll leave it out.

Fedosin commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+  resources:
+    requests:
+  	storage: 10Gi
+  storageClassName: gold <2>
```
All StorageClasses for Manila are provisioned by the Operator and have the `csi-manila-` prefix: https://github.com/openshift/csi-driver-manila-operator/blob/master/pkg/controller/maniladriver/manila_storageclasses.go#L17

Fedosin commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+----
++
+<1> Use RWX if you want the persistent volume (PV) that fulfills this PVC to be mounted to multiple Pods on multiple nodes in the cluster. Use ReadWriteOnce (RWO) mode to prevent additional Pods from being dynamically provisioned.
+<2> The name of the storage class that provisions the storage back end. This was created by the cluster administrator.
```
II don't know if I should go into details, but as I said above, all StorageClasses are automatically provisioned by the Manila CSI driver operator. Cluster administrator can create additional classes, but it doesn't make much sense.

Fedosin commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,89 @@
+. Verify that the storage class was created successfully by entering the following command:
++
+----
+$ oc get storageclass csi-manila-ceph
```
One notice here... StorageClass names can be different, depending on Manila service configuration. 

For instance, in PSI we have just one Manila share type - `ceph`. It means that the operator creates a StorageClass `csi-manila-ceph`. In the upstream they use `nfs`, therefore their StorageClass is called `csi-manila-nfs`: https://github.com/kubernetes/cloud-provider-openstack/blob/master/examples/manila-csi-plugin/nfs/dynamic-provisioning/storageclass.yaml#L4 
And of course there can be several `csi-manila-` StorageClasses, one for each Manila share type.

I think it should be like:

> . Verify that storage classes were created successfully by entering the following command:

and 

> $ oc get storageclasses | grep -E "NAME|csi-manila-"

Fedosin commented on modules/persistent-storage-csi-manila-install-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+
+. Click *Install*.
+
+Once finished, the Manila CSI Driver Operator is listed in the *Installed Operators* section of the web console.
```
btw, I renamed the operator as you asked, from `CSI Driver Manila Operator` to `Manila CSI Driver Operator`. So this part is correct now :)

Fedosin commented on storage/container_storage_interface/persistent-storage-csi-manila.adoc:
```diff
@@ -0,0 +1,24 @@
+Familiarity with link:https://kubernetes.io/docs/concepts/storage/persistent-volumes/[PVs], link:https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims[persistent volume claims (PVCs)], link:http://blog.kubernetes.io/2016/10/dynamic-provisioning-and-storage-in-kubernetes.html[dynamic
+provisioning], and link:https://kubernetes.io/docs/admin/authorization/rbac/[RBAC authorization] is recommended.
+
+Before PVCs can be created, you must install the Manila CSI Driver Operator. The Operator creates the ManilaDriver Custom Resource that is required in the {product-title} cluster. It also creates the required storage classes for all available Manila share types needed for dynamic provisioning.
```
The Operator doesn't create the ManilaDriver Custom Resource. It should be created by the cluster administrator manually after the installation from this manifest: https://github.com/openshift/csi-driver-manila-operator/blob/master/deploy/crds/csi.openshift.io_v1alpha1_maniladriver_cr.yaml

bobfuru replied:
Good catch. I've updated to mention that the driver must be installed manually and linked to the related set of instructions that come later on the page.

bobfuru commented:
Hi @tombarron and @Fedosin - thank you both for your feedback. I've applied changes and responded to your comments in the PR. An updated preview build that incorporates edits is available here:
https://osdocs-930-manila--ocpdocs.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-manila.html

bobfuru commented:
Hi @tombarron and @Fedosin - thank you both for your feedback. I've applied changes and responded to your comments in the PR. An updated preview build that incorporates these edits is available here:
https://osdocs-930-manila--ocpdocs.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-manila.html

**Can you PTAL and can I get a LGTM from you both? Thanks!**

duanwei33 commented:
Hi @bobfuru 
In section "Installing the OpenStack Manila CSI driver"
Please note step 1-4 is install driver from GUI and step 5 is for CLI, they are either-or option, better to make it more clear.
And if using CLI, you can add the yaml file for driver creation.
$ more maniladriver.yaml
apiVersion: csi.openshift.io/v1alpha1
kind: ManilaDriver
metadata:
  name: cluster

bobfuru commented:
Note that there is related work in https://github.com/openshift/openshift-docs/pull/22978. Once that PR is merged, the same guidance should be applied to this PR to use `{rh-openstack-first}` and `{rh-openstack}` attributes.

bobfuru commented:
> Hi @bobfuru
> In section "Installing the OpenStack Manila CSI driver"
> Please note step 1-4 is install driver from GUI and step 5 is for CLI, they are either-or option, better to make it more clear.
> And if using CLI, you can add the yaml file for driver creation.
Thanks @duanwei33 for pointing that out. It makes much more sense now. 😁 
PTAL and let me know if the instructions in this section will work.

duanwei33 commented:
> > In section "Installing the OpenStack Manila CSI driver"
> > Please note step 1-4 is install driver from GUI and step 5 is for CLI, they are either-or option, better to make it more clear.
> > And if using CLI, you can add the yaml file for driver creation.
> 
> Thanks @duanwei33 for pointing that out. It makes much more sense now! 
> PTAL and let me know if the instructions in this section will work.

Thanks for the update, it looks good to me.

Fedosin commented:
Thanks Bob! lgtm!

gnufied commented on storage/container_storage_interface/persistent-storage-csi-manila.adoc:
```diff
@@ -0,0 +1,26 @@
+Before PVCs can be created, you must install the Manila CSI Driver Operator. The Operator creates the required storage classes for all available Manila share types needed for dynamic provisioning.
+
+After the Operator is installed, you must also create the ManilaDriver Custom Resource (CR) that is required in the {product-title} cluster.
+
```
We should cover operator uninstall little bit and add a comment that, it is not recommended to uninstall the operator while there are PVs that are using it.

bobfuru replied:
Thanks, @gnufied - I was wondering about that. I've added a section on uninstalling the operator that is similar to what we are using for AWS EBS operator uninstall: https://github.com/openshift/openshift-docs/pull/22199/files#diff-07baabcc25ef2ac37420e3e62bc64080

bobfuru commented:
> Note that there is related work in #22978. Once that PR is merged, the same guidance should be applied to this PR to use `{rh-openstack-first}` and `{rh-openstack}` attributes.

Done.

tombarron commented on modules/persistent-storage-csi-manila-install-operator.adoc:
```diff
@@ -0,0 +1,32 @@
+
+. On the *Install Operator* page, select *openshift-manila-csi-driver-operator* from the *Installed Namespace* drop-down menu.
+
+. Adjust the values for *Update Channel* and *Approval Strategy* to the values that you want. Only *All namespaces on the cluster* is the only *Installation Mode* that is supported.
```
Nit, but to my ear it sounds better to say "The only supported *Installation Mode* is *All namespaces on the cluster*."

tombarron commented on modules/persistent-storage-csi-manila-dynamic-provisioning.adoc:
```diff
@@ -0,0 +1,87 @@
+{product-title} installs a StorageClass for each available Manila share type.
+
+The YAML files that are created are completely decoupled from Manila and from its Container Storage Interface (CSI) plug-in. As an application developer, you can dynamically provision ReadWriteMany (RWX) storage and deploy Pods with applications that safely consume the storage using YAML manifests.
+
```
Perhaps add "You can also provision other access modes such as ReadWriteOnce (RWO)."   We cover that at line 38, and we want to emphasize RWX here since lots of other drivers don't have support for that mode, but we don't want to give the impression that somehow this driver can *only* do RWX.

bobfuru replied:
Yeah, I wondered whether it might lead to that impression. Will include your addition.

tombarron commented on modules/persistent-storage-csi-manila-install-driver.adoc:
```diff
@@ -0,0 +1,105 @@
++
+----
+NAME              PROVISIONER               RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION  AGE
+csi-manila-ceph   manila.csi.openstack.org  Delete          Immediate           false                 18h
```
Let's substitute 'csi-manila-gold' for 'csi-manila-ceph'.  The 'ceph' here comes from a cut-and-paste of output from PSI cloud where they happen to have a Manila share-type called 'ceph'.  'gold' is more generic and will raise fewer questions (Manila can be backed by Ceph storage but can also be backed by other sorts of storage).

bobfuru commented:
> But I think manila-csi should be included in the section dedicated to dynamic provisioning as well [1].
> It could appear in the "Available dynamic provisioning plug-ins" table right after OpenStack cinder. The provisioner plug-in name is "manila.csi.openstack.org". Notes could just be a link back to the manila-csi docs that you are currently writing. There should be a "RHOSP Manila CSI object definition" section that says the Manila CSI storage classes are automatically created when the Operator and Driver are installed as covered in [link](to this section you are writing now).

Great idea! I've added CSI Manila as you requested. I'm not 100% I understand what you mean by an "object definition section" or where it should go. I've included in the table notes that installation of the CSI Manila Operator/driver creates the storage classes for dynamic provisioning. Hopefully, this inclusion addresses your request in full: https://github.com/openshift/openshift-docs/pull/22199/files#diff-314d980bf1d6c442ecf9c1d0b8ed15feR24-R26

tombarron commented:
LGTM, Bob.  Thanks for your work on this one!

bobfuru commented:
/cherrypick enterprise-4.5
